### PR TITLE
[SPARK-49111][SQL] Move withProjectAndFilter to the companion object of DataSourceV2Strategy

### PR DESCRIPTION
**Pull request by urosstan-db**
### What changes were proposed in this pull request?
Move static method `withProjectAndFilter` to object in DataSourceV2Strategy

### Why are the changes needed?
It provides better opportunities for reuse, since object of strategy is not needed anymore for function invocation, and a code is also cleaner. 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Simple refactor, there is no new changes that can be tested.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Review by PetarVasiljevic-DB — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala [hunk 1]
```diff
@@ -645,6 +633,21 @@ private[sql] object DataSourceV2Strategy extends Logging {
       logWarning(log"Can't translate ${MDC(EXPR, other)} to source filter, unsupported expression")
       None
   }
+
+  private def withProjectAndFilter(
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:637]: Please write a function description like https://github.com/apache/spark/pull/47606/files#diff-7aeb491d44e183c8c8cf86d90b57701dba009fc19983c2a5c09449c768b047ceR549-R557 ?

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:637]: Should we change it to `protected[sql]` like the other utility function in DataSourceV2Strategy?

Comment (urosstan-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala]: Makes sense, I addressed both comments

**Review by MaxGekk — COMMENTED** [review]
@urosstan-db Please, add the tag `[SQL]`, and make the PR's title more concrete like "Move `withProjectAndFilter` to the companion object of `DataSourceV2Strategy`" otherwise LGTM

**Comment by urosstan-db** [comment]
> @urosstan-db Please, add the tag `[SQL]`, and make the PR's title more concrete like "Move `withProjectAndFilter` to the companion object of `DataSourceV2Strategy`" otherwise LGTM

Thank you, done

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @urosstan-db and @gatorsmile @PetarVasiljevic-DB for review.

**Comment by MaxGekk** [comment]
+1, LGTM. Merged to master.
Thank you, @urosstan-db and @gatorsmile @PetarVasiljevic-DB for review.

**Pull request closed without merging**